### ISSUE TITLE
[Reverse Proxy] Communication through Azure Application Gateway blocked for WebSocket traffic

### ISSUE THREAD

**Issue by alecor191** [Azure/service-fabric-issues#421]
[Updated: Performed tests also using HTTP/WS and updated table below with results]

I'm observing an issue where WebSocket traffic over TLS (wss://) doesn't flow between AppGateway and SF Reverse Proxy. The scenario is as follows (the red arrow shows where traffic doesn't flow)

![flow](https://user-images.githubusercontent.com/6270008/29576989-1d998bba-876a-11e7-9b18-ece8d5728d03.png)

Please note that:
- *HTTPS* traffic works fine. I.e. in the scenario above, HTTPS traffic to the same SF service works fine.
- WSS traffic to the SF LB (that calls SF RP, which then calls our service) works fine
- WSS traffic to our service through AppGateway works (i.e. skip SF RP and let AppGateway connect directly to our service)
- Please note that the initial wss:// upgrade *request reaches our service* (i.e. incoming request gets through). Our service then replies with a `101 Upgrade`. However, this *response never reaches to the browser*.

To summarize, here the combinations we tried:

| Scenario | HTTPS | WSS | HTTP | WS |
|---|---|---|---|---|
| **AppGateway** + SF LB + **SF Reverse Proxy** + Our service | works | **Doesn't work** | works | **Doesn't work** |
| SF LB + **SF Reverse Proxy** + Our service | works | works | works | works |
| **AppGateway** + SF LB + Our service | works | works | works | works |

As you can see all combinations work, besides the one we're interested in: Having AppGateway call SF RP for WSS traffic.

Any idea how to fix this?

**Comment by kavyako** [comment]
This is a bug in the ApplicationGateway, tentative fix ~June 2018.

Details:
When ApplicationGateway forwards the request to the reverse proxy, it clears the Sec-Websocket-Extension header and sends an empty value for the header.
As per RFC, the Sec-Websocket-Extension header must have atleast one value. Empty implies malformed. Hence Service Fabric reverse proxy fails the handshake.

I would recommend opening an issue with the ApplicationGateway repo for tracking purposes.

**Comment by alecor191** [comment]
Thanks a lot @kavyako for the additional details! 

You mention a tentative fix date of 6/18 but also that I should open a bug in AppGateway repo. To understand: does this mean that there is already a bug open with AppGateway team and they're working on it? 

Could you please point me to the AppGateway repo where I can open the issue? (I searched GitHub but couldn't find one with a matching name)

**Comment by FesterCluck** [comment]
@kavyako @alecor191 the AppGateway team is aware of the issue, and has a few bugs open on the matter. From their response I concluded an implementation supporting permessage-deflate was unlikely. Is this still the case? It was regarding routing.

**Comment by kavyako** [comment]
@alecor191 Looks like AppGateway team does not have a public github forum. Their suggestion is to open a feature request via tam/csa etc or open a uservoice forum request for your tracking purposes.

**Comment by kavyako** [comment]
Closing this issue since it is not a reverse proxy issue.
